File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;

public class Main{
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		String[] line = br.readLine().trim().split(" ");
		int K = Integer.parseInt(line[0]);
		int N = Integer.parseInt(line[1]);
		long[] cables = new long[K];
		long high = 0;
		long low = 1;
		long result = 0;
		long mid = 0;
		
		for(int i = 0; i < K; i++){
            cables[i] = Long.parseLong(br.readLine().trim());
			if (high < cables[i]) high = cables[i];
		}
		
		while(low <= high){
            long count = 0;
			mid = (low + high) / 2;
			for(long cable : cables){
                count += cable / mid;
            }

			if(count >= N){
				result = mid;
                low = mid + 1;
			}else{
				high = mid - 1;
			}
		}
		
		bw.write(String.valueOf(result) + '\n');
		bw.flush();
		bw.close();
		br.close();
	}
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
import sys

input = sys.stdin.readline

K, N = map(int, input().strip().split())

cables = list(int(input().strip()) for _ in range(K))

high = max(cables)
low = 1
result = 0

while low <= high:
    count = 0
    mid = (low + high) // 2

    for cable in cables:
        count += cable // mid

    if count >= N:
        result = mid
        low = mid + 1
    else:
        high = mid - 1

sys.stdout.write(str(result) + '\n')